cpp/src/gandiva/precompiled/string_ops.cc

```diff
@@ -2837,21 +2840,30 @@ const char* soundex_utf8(gdv_int64 ctx, const char* in, gdv_int32 in_len,
   for (int i = start_idx, l = in_len; i < l; i++) {
     if (isalpha(in[i]) > 0) {
       c = toupper(in[i]) - 65;
-      if (mappings[c] != '0') {
-        if (mappings[c] != ret[si - 1]) {
-          ret[si] = mappings[c];
-          si++;
-        }
-
-        if (si > 3) break;
+      if (mappings[c] != soundex[si - 1]) {
+        soundex[si] = mappings[c];
+        si++;
       }
     }
   }
 
-  if (si <= 3) {
-    while (si <= 3) {
-      ret[si] = '0';
-      si++;
+  for (int i = 1; i < si; i++) {
+    // If the saved letter's digit is the same as the resulting first digit, remove the
+    // digit.
+    if (i == 1 && soundex[i] == mappings[ret[0] - 65]) {
+      i++;
+    }
+    if (soundex[i] != '0') {
```

Comment:
I see this skipping all vowels and anything that maps to '0'. This is step 1 in the algorithm in line 2792.

Are we implementing step 3 as well?

3. If two or more letters with the same number are adjacent in the original name
//    (before step 1), only retain the first letter; 

I can see this implemented in line 2843. However, I dont see the below implemented. If we are not implementing the below, please remove this from the comments

also two letters with the same number
//    separated by 'h' or 'w' are coded as a single number, whereas such letters separated
//    by a vowel are coded twice. This rule also applies to the first letter.

Reply:
I updated the documentation, look if is good